File: core/src/main/java/com/cmhteixeira/bittorrent2/tracker/BadTracker.java
package com.cmhteixeira.bittorrent2.tracker;

import com.cmhteixeira.bittorrent.InfoHash;
import com.cmhteixeira.bittorrent.UdpSocket;
import com.google.common.collect.ImmutableList;
import com.google.common.collect.ImmutableMap;
import com.google.common.collect.ImmutableSet;

import java.io.IOException;
import java.net.DatagramPacket;
import java.net.DatagramSocket;
import java.net.InetSocketAddress;
import java.util.Map;

public class BadTracker implements TrackerJava {
  ImmutableMap<InetSocketAddress, DatagramSocket> sockets;
  final int packetSize = 65507;

  public BadTracker(ImmutableMap<InetSocketAddress, DatagramSocket> sockets) {
    this.sockets = sockets;
    Thread thread = new Thread(new ReadMessagesFromSockets(), "ProcessingThread");
    thread.start();
  }

  @Override
  public void submit(InfoHash torrent, ImmutableList<UdpSocket> trackers) {}

  @Override
  public ImmutableSet<InetSocketAddress> peers(InfoHash torrent) {
    return null;
  }

  private class ReadMessagesFromSockets implements Runnable {

    @Override
    public void run() {
      while (true) {
        for (Map.Entry<InetSocketAddress, DatagramSocket> tracker : sockets.entrySet()) {
          DatagramPacket packet = new DatagramPacket(new byte[packetSize], packetSize);
          InetSocketAddress trackerSocketAddress = tracker.getKey();
          DatagramSocket trackerSocket = tracker.getValue();
          try {
            trackerSocket.receive(packet);
          } catch (IOException e) {
            throw new RuntimeException(e);
          }
          //          process(packet);
        }
      }
    }
  }
}

File: core/src/main/scala/com/cmhteixeira/bittorrent/tracker/BadTracker.scala
package com.cmhteixeira.bittorrent.tracker
import com.cmhteixeira.bittorrent.InfoHash

import java.net.{DatagramPacket, DatagramSocket, InetSocketAddress}
import java.nio.ByteBuffer

class BadTracker private(sockets: Map[InetSocketAddress, DatagramSocket]) extends Tracker {
  override def peers(infoHash: InfoHash): Set[InetSocketAddress] = ???
  override def statistics: Map[InfoHash, Tracker.Statistics] =
    ???
  override def submit(torrent: Torrent): Unit = ???
}

object BadTracker {
  private val maximumUdpPacketSize = 65507

  private class ReadMessagesFromSockets(sockets: Map[InetSocketAddress, DatagramSocket]) extends Runnable {
    override def run(): Unit = {
      sockets.foreach { case (socketAddress, socket) =>
        val packet = new DatagramPacket(ByteBuffer.allocate(maximumUdpPacketSize).array(), maximumUdpPacketSize)
        socket.receive(packet)
        processPacket(packet)
      }
      run()
    }
    private def processPacket(packet: DatagramPacket) = ???
  }

  def apply(sockets: Map[InetSocketAddress, DatagramSocket]): BadTracker = {
    val thread = new Thread(new ReadMessagesFromSockets(sockets), "ProcessingThread")
    thread.start()
    new BadTracker(sockets)
  }
}
